data/downloader/downloader.go

```diff
@@ -131,20 +131,21 @@ func (d *Downloader) handleImport(item *DownloadItem) {
 			d.failedQueue[item.URI] = item
 			d.failedQueueLock.Unlock()
 		} else {
-			log.Warnw("could not pin file", "uri", item.URI, "error", fmt.Sprintf("%v", err))
+			log.Warnw("could not retrieve file", "uri", item.URI, "error", fmt.Sprintf("%v", err))
 		}
 		return
 	}
-	ctx, cancel = context.WithTimeout(context.Background(), ImportPinTimeout)
-	defer cancel()
-	data, err := d.RemoteStorage.Retrieve(ctx, item.URI, 0)
-	if err != nil {
-		log.Warnw("could not retrieve file", "uri", item.URI, "error", fmt.Sprintf("%v", err))
-	} else {
-		if item.Callback != nil {
-			go item.Callback(item.URI, data)
+	go func() {
+		ctx, cancel := context.WithTimeout(context.Background(), ImportPinTimeout)
+		defer cancel()
+		if err := d.RemoteStorage.Pin(ctx, item.URI); err != nil {
+			log.Warnw("could not pin file", "uri", item.URI, "error", fmt.Sprintf("%v", err))
 		}
+	}()
+	if item.Callback != nil {
+		go item.Callback(item.URI, data)
```

Comment:
IMO callbacks should never be asynchronous. if the callback needs to do anything async, it can fire a goroutine on its own.